### PR TITLE
search router polish

### PR DESCRIPTION
**Pull request by 289Adam289**
### Details

This pr fixes following problems mentioned in https://github.com/Expensify/App/issues/50250:
 - parsing error with special characters
 - amount display value in cents
 - improve performance of filter validation
 - fix reordering of standard filters
 - [Using the router ](https://expensify.slack.com/archives/C06ML6X0W9L/p1728287879352319?thread_ts=1727907702.678009&cid=C06ML6X0W9L)[ and putting a link removes the link for the search results](https://expensify.slack.com/archives/C06ML6X0W9L/p1728287879352319?thread_ts=1727907702.678009&cid=C06ML6X0W9L)
### Fixed Issues

https://github.com/Expensify/App/issues/50250
https://github.com/Expensify/App/issues/50988
PROPOSAL:

### Tests

- Open Search Page
- Enter custom search query with `amount`
- Verify that filters except default filters and free text don't change order
- Verify that `amount` filter is not displayed in cents 
- Verify that special characters e.g. emoji can be used in search input

- [x] Verify that no errors appear in the JS console

### QA Steps

- Open Search Page
- Enter custom search query with `amount`
- Verify that filters except default filters and free text don't change order
- Verify that `amount` filter is not displayed in cents 
- Verify that special characters e.g. emoji can be used in search input

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9924d6ff-a062-409a-8ac6-a8754440c4be
</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/69e28171-c633-46e9-8df6-fd15632eca0e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/3a7b25f8-f9b1-46b4-bc4b-e1597054a12e

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/7d7c333d-722a-402c-a6d7-38377d8f8ef8
</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Review by SzymczakJ — APPROVED** [review]
LGTM 🚀

**Comment by 289Adam289** [comment]
@rayane-djouah I am not sure why but no reviewer has been assigned to this pr. You have reviewed https://github.com/Expensify/App/pull/49838 and I think you can review this one as a follow up pr to https://github.com/Expensify/App/pull/49838.

**Review by rayane-d — COMMENTED** [review]
.

File: src/libs/SearchUtils.ts [hunk 1]
```diff
@@ -726,12 +723,16 @@ function buildFilterFormValuesFromQuery(
...
-            filtersForm[FILTER_KEYS.DATE_AFTER] = filters[filterKey]?.find((filter) => filter.operator === 'gt' && ValidationUtils.isValidDate(filter.value.toString()))?.value.toString();
+            filtersForm[FILTER_KEYS.DATE_BEFORE] = filterList.find((filter) => filter.operator === 'lt' && ValidationUtils.isValidDate(filter.value.toString()))?.value.toString();
+            filtersForm[FILTER_KEYS.DATE_AFTER] = filterList.find((filter) => filter.operator === 'gt' && ValidationUtils.isValidDate(filter.value.toString()))?.value.toString();
         }
         if (filterKey === CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT) {
-            filtersForm[FILTER_KEYS.LESS_THAN] = filters[filterKey]?.find((filter) => filter.operator === 'lt' && validateAmount(filter.value.toString(), 2))?.value.toString();
-            filtersForm[FILTER_KEYS.GREATER_THAN] = filters[filterKey]?.find((filter) => filter.operator === 'gt' && validateAmount(filter.value.toString(), 2))?.value.toString();
+            filtersForm[FILTER_KEYS.LESS_THAN] = filterList
+                .find((filter) => filter.operator === 'lt' && validateAmount(filter.value.toString(), 0, CONST.IOU.AMOUNT_MAX_LENGTH + 2))
```
Comment (luacmartins) [inline src/libs/SearchUtils.ts:731]: Maybe we can add a comment on why we're using `CONST.IOU.AMOUNT_MAX_LENGTH + 2`

**Review by luacmartins — COMMENTED** [review]
Nice changes. Was the link dropping issue a parsing problem?

**Comment by 289Adam289** [comment]
Yes, the link dropping issue was a parsing problem. When you pasted the link parsing error occurred and router navigated to search page with the last query without error.

**Comment by luacmartins** [comment]
@289Adam289 we have conflicts and some tests are failing

**Review by luacmartins — COMMENTED** [review]
@289Adam289 we have conflicts and some tests are failing

File: src/libs/SearchParser/baseRules.peggy [hunk 2]
```diff
@@ -19,10 +19,10 @@ operator "operator"
...
 alphanumeric "word"
-  = chars:[A-Za-z0-9_@./#&+\-\\',;%]+ {
+  = chars:[^ "\t\n\r]+ {
       return chars.join("").trim().split(",").filter(Boolean);
     }
 
 logicalAnd = _ { return "and"; }
 
-_ "whitespace" = [ \t\r\n]*
+_ "whitespace" = [ \t\r\n]*
```
Comment (luacmartins) [inline src/libs/SearchParser/baseRules.peggy:28]: NAB add empty line at end of file

**Review by luacmartins — DISMISSED** [review]
LGTM and tests well. @rayane-djouah let's prioritize reviewing this one

**Review by rayane-d — CHANGES_REQUESTED** [review]

**Comment by luacmartins** [comment]
We also have conflicts now

**Comment by 289Adam289** [comment]
@rayane-djouah I fixed both bugs and resolved conflicts

**Review by luacmartins — DISMISSED** [review]

**Comment by luacmartins** [comment]
all yours @rayane-djouah

**Comment by 289Adam289** [comment]
I gave it a try and I think it works. Unit tests pass but further testing is needed because I've heavily changed important rules. cc @rayane-djouah

**Review by luacmartins — APPROVED** [review]
LGTM. I did some testing on the latest grammar and it seems to be working well. All yours @rayane-djouah

**Review by rayane-d — APPROVED** [review]
Works well

**Comment by dangrous** [comment]
Hi! I haven't confirmed yet but it seems like this is related to https://github.com/Expensify/App/issues/51661 - can you take a quick look?

File: src/libs/SearchQueryUtils.ts [hunk 3]
```diff
@@ -218,18 +226,14 @@ function getQueryHash(query: SearchQueryJSON): number {
     orderedQuery += ` ${CONST.SEARCH.SYNTAX_ROOT_KEYS.SORT_BY}:${query.sortBy}`;
     orderedQuery += ` ${CONST.SEARCH.SYNTAX_ROOT_KEYS.SORT_ORDER}:${query.sortOrder}`;
 
-    Object.keys(query.flatFilters)
+    query.flatFilters.forEach((filter) => {
+        filter.filters.sort((a, b) => localeCompare(a.value.toString(), b.value.toString()));
```
Comment (shubham1206agra) [inline src/libs/SearchQueryUtils.ts:230]: Mark for ref

File: src/libs/SearchQueryUtils.ts [hunk 4]
```diff
@@ -200,6 +198,16 @@ function findIDFromDisplayValue(filterName: ValueOf<typeof CONST.SEARCH.SYNTAX_F
             )
             .flat();
     }
+    if (filterName === CONST.SEARCH.SYNTAX_FILTER_KEYS.AMOUNT) {
+        if (typeof filter === 'string') {
+            const backendAmount = CurrencyUtils.convertToBackendAmount(Number(filter));
```
Comment (DylanDylann) [inline src/libs/SearchQueryUtils.ts:203]: This change caused https://github.com/Expensify/App/issues/79934, 


> This PR added amount-to-cents conversion (convertToBackendAmount) to the findIDFromDisplayValue function in src/libs/SearchQueryUtils.ts. The PR's intent was to fix "amount display value in cents," but this function was already being called for all search submissions from SearchRouter.tsx — including clicks on recent search items.

**Pull request merged**